src/appleseed/foundation/image/filteredtile.cpp

```diff
@@ -163,4 +165,57 @@ void FilteredTile::atomic_add(
     }
 }
 
+float FilteredTile::compute_weighted_pixel_variance(
+    const float*        main,
+    const float*        second)
+{
+    // Get weights.
+    float main_weight = *main++;
+    main_weight = main_weight == 0.0f ? 0.0f : 1.0f / main_weight;
+    float second_weight = *second++;
+    second_weight = second_weight == 0.0f ? 0.0f : 1.0f / second_weight;
+
+    // Get colors and assign weights.
+    Color4f main_color(abs(main[0]), abs(main[1]), abs(main[2]), abs(main[3]));
+    main_color *= main_weight;
+
+    Color4f second_color(abs(second[0]), abs(second[1]), abs(second[2]), abs(second[3]));
+    second_color *= second_weight;
+
+    // Compute variance.
+    return (
+        abs(main_color.r - second_color.r)
+        + abs(main_color.g - second_color.g)
+        + abs(main_color.b - second_color.b)
+       ) / fast_sqrt(main_color.r + main_color.g + main_color.b);
+}
+
+float FilteredTile::compute_tile_variance(
+    const AABB2u&       bb,
+    FilteredTile*       main,
+    FilteredTile*       second)
+{
+    float error = 0.0;
+
+    assert(main->get_crop_window() == second->get_crop_window());
+    assert(main->get_crop_window().contains(bb.min));
+    assert(main->get_crop_window().contains(bb.max));
+    assert(bb.min.y >= 0 && bb.max.y <= main->get_height());
+    assert(bb.min.x >= 0 && bb.max.x <= main->get_width());
```

Comment:
Switch these two assertions (x first, y second). Are they actually necessary given the two assertions above?

Reply:
No :grimacing: